x/wasm/keeper/keeper.go

```diff
@@ -1213,3 +1218,68 @@ func (h DefaultWasmVMContractResponseHandler) Handle(ctx sdk.Context, contractAd
 	}
 	return result, nil
 }
+
+// GetByteCodeByChecksum queries the wasm code by checksum
+func (k Keeper) GetByteCodeByChecksum(ctx sdk.Context, checksum wasmvm.Checksum) ([]byte, error) {
+	return k.wasmVM.GetCode(checksum)
+}
+
+// iteratePinnedCodes iterates over all pinned code ids in ascending order
+func (k Keeper) iteratePinnedCodes(ctx sdk.Context, cb func(codeID uint64) bool) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.PinnedCodeIndexPrefix)
+	iter := store.Iterator(nil, nil)
+	defer iter.Close()
+
+	for ; iter.Valid(); iter.Next() {
+		codeID := sdk.BigEndianToUint64(iter.Key())
+		if cb(codeID) {
+			return
+		}
+	}
+}
+
+// PruneWasmCodes deletes code info for unpinned codes <= than latestCodeID
+func (k Keeper) PruneWasmCodes(ctx sdk.Context, latestCodeID uint64) error {
+	usedCodeIds := make(map[uint64]struct{}, 0)
+	usedChecksums := make(map[string]bool, 0)
+
+	// collect code ids used by contracts
+	k.IterateContractInfo(ctx, func(_ sdk.AccAddress, info types.ContractInfo) bool {
+		usedCodeIds[info.CodeID] = struct{}{}
+		return false
+	})
+
+	// collect pinned code ids
+	k.iteratePinnedCodes(ctx, func(codeID uint64) bool {
+		usedCodeIds[codeID] = struct{}{}
+		return false
+	})
+
+	// check if instances are used only by unpinned code ids <= latestCodeID
+	k.IterateCodeInfos(ctx, func(codeID uint64, info types.CodeInfo) bool {
+		if _, ok := usedCodeIds[codeID]; !ok && codeID <= latestCodeID {
+			if _, found := usedChecksums[string(info.CodeHash)]; !found {
+				usedChecksums[string(info.CodeHash)] = false
+			}
+			return false
+		}
+		usedChecksums[string(info.CodeHash)] = true
+		return false
+	})
+
+	// delete all unpinned code ids <= latestCodeID and all the
+	// instances used only by unpinned code ids <= latestCodeID
+	k.IterateCodeInfos(ctx, func(codeID uint64, info types.CodeInfo) bool {
+		if _, ok := usedCodeIds[codeID]; !ok && codeID <= latestCodeID {
```

Comment:
The conditions make sense here. I would prefer some better naming for `latestCodeID`. Maybe `maxCodeID`? Latest is a moving target for me as there can be new code-infos stored while the gov proposal is in-flight